File: runtime/ruby/units/wasi/path_unlink_file.rb
# requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
def wasi_path_unlink_file(dirfd, path_ptr, path_len)
  rel = @memory.read_string(path_ptr, path_len)
  # unlink(2) never follows a trailing symlink: it removes the link.
  host_path, err = resolve_path(dirfd, rel, follow_last: false)
  return err if err
  # A slash-suffixed non-directory is ENOTDIR, missing is ENOENT; a real
  # directory falls through to File.unlink (host EPERM/EISDIR). Probe the
  # slash-stripped path: stat on "file/" fails ENOTDIR, reads as missing.
  if host_path.end_with?("/")
    bare = host_path.delete_suffix("/")
    unless File.directory?(bare)
      return File.exist?(bare) || File.symlink?(bare) ? ERRNO_NOTDIR : ERRNO_NOENT
    end
  end
  File.unlink(host_path)
  ERRNO_SUCCESS
rescue SystemCallError => e
  fs_errno(e)
end

File: runtime/go/units/wasi/path_unlink_file.go
// requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
func (w *WASI) wasi_path_unlink_file(dirfd, pathPtr, pathLen uint32) uint32 {
    rel := string(w.memory.read_string(uint64(pathPtr), uint64(pathLen)))
    // unlink(2) never follows a trailing symlink: it removes the link. The raw
    // syscall fails (EISDIR/EPERM) on a directory rather than removing it.
    hostPath, err := w.resolve_path(dirfd, rel, false)
    if err != wasiOk {
        return err
    }
    if e := syscall.Unlink(hostPath); e != nil {
        return w.fs_errno(e)
    }
    return wasiOk
}
